contributors/design-proposals/pod-injection-policy.md

```diff
@@ -0,0 +1,447 @@
+    ObjectMeta
+
+    Spec PodInjectionPolicySpec
+}
+
+type PodInjectionPolicySpec struct {
+    // Selector defines which Pods should be modified, this only targets Pods.
+    Selector      unversioned.LabelSelector
+    // Env defines the collection of EnvVar to inject.
+    Env           []EnvVar `json:omitempty`
+    // EnvFrom defines the collection of EnvFromSource to inject.
+    EnvFrom       []EnvFromSource `json:omitempty`
+    // Volumes defines the collection of Volume to inject.
+    Volumes       []Volume `json:omitempty`
+    // VolumeMounts defines the collection of VolumeMount to inject.
+    VolumeMounts  []VolumeMount `json:omitempty`
+}
+```
+
+#### Validations
+
+In order for the Pod Injection Policy to be valid it must fulfill the
+following constraints:
+
+- The `Selector` field must be defined. This is how we know which pod templates
+  to inject so therefore it is required.
+- The policy must define _at least_ 1 of `Env`, `EnvFrom`, `Volumes`, or
+  `VolumeMounts`.
+- The `Volume`s given in `Volumes` must exist and be valid.
+
+#### Conflicts
+
+There are a number of edge conditions that might occur at the time of
+injection. These are as follows:
+
+- Merging lists with no conflicts: if a pod already has a volume defined
+  exactly as defined in the SIP. This would replace the other volume.
+- Merging lists with conflicts: if a SIP redefines an `EnvVar` or a `Volume`,
+  the original will be replaced and an annotation to the Pod spec will reflect
+  a warning documenting the decision.
+- Conflicts between `Env` and `EnvFrom`: this would throw an error.
+
+### AdmissionControl Plug-in: PodInjectionPolicy
+
+The **PodInjectionPolicy** plug-in introspects all incoming pod creation
+requests and injects the pod based off a `Selector` with the desired
+attributes.
+
+For the initial alpha, the order of precedence for applying multiple
+`PodInjectionPolicy` specs is from oldest to newest. All Pod Injection
+Policies in a namespace should be order agnostic. However we can use merge keys
+to detect some of the conflicts that may occur.
+
+This will not be enabled by default for all clusters, but once GA will be
+a part of the set of strongly recommended plug-ins documented
+[here](https://kubernetes.io/docs/admin/admission-controllers/#is-there-a-recommended-set-of-plug-ins-to-use).
+
+**Why not an Initializer?**
+
+This will be first implemented as an AdmissionControl plug-in then can be
+converted to an Initializer once that is fully ready.
+
+
+#### Behavior
+
+This will modify the pod spec. The supported changes to
+`Env`, `EnvFrom`, and `VolumeMounts` apply to the container spec for
+all containers in the pod with the specified matching `Selector`. The
+changes to `Volumes` apply to the pod spec for all pods matching `Selector`.
+
+The resultant modified pod spec will be annotated to show that it was modified by
+the `PodInjectionPolicy`. This will be of the form
+`podinjectionpolicy.k8s.io": "<sip name>"`.
```

Comment:
First, this doesn't follow the proposed conventions:
https://github.com/kubernetes/community/issues/300

Second, @philips made a good point about multiple injection policies. An annotation key of the form `<annotation prefix>/<pip name>` would enable that.